src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/DotnetCliHelper.cs

```diff
@@ -0,0 +1,172 @@
+    {
+        _logger.LogDebug($"Running dotnet CLI command at {_dotnetExecutablePath.Value} in directory {workingDirectory} with arguments {arguments}");
+
+        var startInfo = new ProcessStartInfo(_dotnetExecutablePath.Value, arguments)
+        {
+            CreateNoWindow = true,
+            UseShellExecute = false,
+            RedirectStandardOutput = true,
+            RedirectStandardError = true
+        };
+
+        if (workingDirectory != null)
+        {
+            startInfo.WorkingDirectory = workingDirectory;
+        }
+
+        // If we're relying on the user's configured CLI to run this we need to remove our custom DOTNET_ROOT that we set
+        // Wto start the server; otherwise we won't find the users installed sdks.
+        if (_dotnetExecutablePath.Value == "dotnet")
+        {
+            startInfo.Environment[DotnetRootEnvVar] = GetUserDotnetRoot();
+        }
+
+        if (!shouldLocalizeOutput)
+        {
+            // The caller requested that we not localize the output of the command (typically be cause it needs to be parsed)
+            // Set the appropriate environment variable for the process so that we don't get localized output.
+            startInfo.Environment["DOTNET_CLI_UI_LANGUAGE"] = "en-US";
+        }
+
+        // Since we depend on MSBuild APIs, the following environment variables get set to the version of dotnet that runs this server.
+        // However want to use the user specified dotnet version to run the tests, so we need to unset these.
+        startInfo.Environment.Remove("MSBUILD_EXE_PATH");
+        startInfo.Environment.Remove("MSBuildExtensionsPath");
+
+        var process = Process.Start(startInfo);
+        Contract.ThrowIfNull(process, $"Unable to start dotnet CLI at {_dotnetExecutablePath.Value} with arguments {arguments} in directory {workingDirectory}");
+        return process;
+    }
+
+    /// <summary>
+    /// Since we change DOTNET_ROOT in order to start the process, we need to get the original value
+    /// whenever we shell out to the CLI or other processes (e.g. vstest console).
+    /// </summary>
+    /// <returns></returns>
+    public static string GetUserDotnetRoot()
+    {
+        // If the user had a valid dotnet root, set it to that, otherwise leave it empty.
+        var dotnetRootUser = Environment.GetEnvironmentVariable("DOTNET_ROOT_USER");
+        var newDotnetRootValue = Path.Exists(dotnetRootUser) ? dotnetRootUser : string.Empty;
+        return newDotnetRootValue;
+    }
+
+    public async Task<string> GetVsTestConsolePathAsync(CancellationToken cancellationToken)
+    {
+        var dotnetSdkFolder = await _dotnetSdkFolder.GetValueAsync(cancellationToken);
+        var vstestConsole = Path.Combine(dotnetSdkFolder, "vstest.console.dll");
+        Contract.ThrowIfFalse(File.Exists(vstestConsole), $"VSTestConsole was not found at {vstestConsole}");
```

Comment:
is throwing the right thing to do here?  are we certain there is always a vstest.console.dll in the dotnet sdk 100% of the time? if not, we should gracefully fail out here stating we couldn't find this dll.

Reply:
Yeah throwing is correct here.  The vstest.console.dll should exist in any SDK we support.